### ISSUE TITLE
[HOLD for payment 2022-11-02] [$2000] Render Emoji when typed with emoji code

### ISSUE THREAD

**Issue by mananjadhav** [Expensify/App#4324]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to the chat
2. Enter `:smile:`

## Expected Result:
- It should've rendered a smiley. Reference screencast from Slack.
https://user-images.githubusercontent.com/3069065/127681715-a265df2f-2a63-4b64-a45a-f577edbbd600.mp4

## Actual Result:
- Video from the current version
https://user-images.githubusercontent.com/3069065/127681725-dbc4786e-44b4-4fcb-b859-858f929ff887.mp4

## Workaround:
- Either stick with the text codes
- Or choose from the Emoji Picker

## Platform:

Where is this issue occurring?

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation
**Expensify/Expensify Issue URL:**

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

**Proposed Solution**
*Update App/assets/emojis.js with the relevant keywords and implement a regex to filter and render the emoji*

**Comment by MelvinBot** [comment]
Triggered auto assignment to @NicMendonca (`AutoAssignerTriage`), see https://stackoverflow.com/c/expensify/questions/4749 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @sketchydroide (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by sketchydroide** [comment]
this in principle sounds like a good idea, but I'm not sure if we already have plans to do this plus the `:smile:` template as well, @roryabraham or @marcaaron do you have any idea if we have plans to get emoji to work like this as part of a bigger project?

**Comment by marcaaron** [comment]
Sorry, I do not

**Comment by roryabraham** [comment]
> do you have any idea if we have plans to get emoji to work like this as part of a bigger project?

Not that I know of, but again I'm not sure. Wbu @stitesExpensify?

**Comment by stitesExpensify** [comment]
We do not have plans to do `:)` or `:p`. It would be easy to add another regex filter once we have things like `:smile:` implemented though.

**Comment by mananjadhav** [comment]
Noted. Are you planning to add a picker where you could :smile: which is currently supported in Github comments and Slack input too?

**Comment by stitesExpensify** [comment]
Yes, we want to do `:smile:` style emojis but have not come up with a solid plan on how yet. 

The problem is that right now we have keywords with overlap. We need to have a default keyword for each emoji that does not overlap. We would like to use the same or similar default keywords as slack, but haven't found a good way to add that to our emoji.js list. Any suggestions?

**Comment by sketchydroide** [comment]
Is this a 2 step thing @stitesExpensify? Step 1 replace the keywords we want to use for emojis that overlap with something else, sept 2 implement the keyword replacement for emojis?

**Comment by stitesExpensify** [comment]
It can be either. Depends on if someone wants to take the whole thing on or just do part.

To clarify though, we are not _replacing_ keywords that overlap as those are still useful for searching. We need to add a _new_ field that has a unique keyword for in-chat replacement

**Comment by mananjadhav** [comment]
We could use this to create the keywords. 

https://github.com/missive/emoji-mart/blob/master/data/all.json We can match the keys by Unicode value.

We can even extend the data source with the above dataset. Currently, we have ~1360 emojis, whereas the above link has ~1740

**Comment by sketchydroide** [comment]
I don't feel this is a daily, as there is no pressing need for this change.
@stitesExpensify do you mind I assign you, it seems you are on top of this way more than me.

**Comment by stitesExpensify** [comment]
Sure, go for it!

**Comment by stitesExpensify** [comment]
@mananjadhav so you're saying we should swap our whole `emoji.js` list for that one?

**Comment by mananjadhav** [comment]
Either swap or at least get the keywords/unique ids for all the emojis based on Unicode.

**Comment by stitesExpensify** [comment]
Personally I'd like to just put uniqueIDs into our current emoji file rather than replace the whole thing

**Comment by mananjadhav** [comment]
and as I mentioned earlier we can do that too.

```
'use strict';

const emojiMap = require('./expensify-emoji.json');
const externalEmojiMap = require('./emoji-mart.json').emojis;
const fs = require('fs');

const emojiTitleMap = fetchEmojiMartMap();

for(let category in emojiMap) {
  let emojis = emojiMap[category];
  for (var i=0; i<emojis.length; i++) {
    // Existing code logic from emoji.js stays here
    // ...
    // ...
    // ...

    // Pick title
    if(emojiTitleMap[emojis[i].code]) {
      emojis[i].title = emojiTitleMap[emojis[i].code].title;

      // Bonus: We can extend keywords too
      emojis[i].keywords = emojis[i].keywords.concat(emojiTitleMap[emojis[i].code].keywords).filter((v, i, a) => a.indexOf(v) === i)
    }
  }

}


function fetchEmojiMartMap () {
  let emojiMap = {};
  for(let emojiKey in externalEmojiMap) {
    const currentEmoji = externalEmojiMap[emojiKey];
    let emojiUnicode = `U+${currentEmoji['b']}`;
    emojiMap[emojiUnicode] = {...currentEmoji, title: emojiKey, keywords: currentEmoji['j']};
  }
  return emojiMap;
}
```

**Comment by stitesExpensify** [comment]
No update, focused on internal priorities

**Comment by stitesExpensify** [comment]
Hi @mananjadhav As part of our internal process, for big projects like this we need to send out changes in a design doc format for internal engineering review. I'm going to be hopefully sending that out soon, but in the mean time i'm demoting this to monthly

**Comment by mananjadhav** [comment]
Thanks for the update @stitesExpensify. Then I think we should also put https://github.com/Expensify/App/issues/4888 on hold as well. Let me know what you think?

**Comment by stitesExpensify** [comment]
This is still holding

**Comment by stitesExpensify** [comment]
No update

**Comment by stitesExpensify** [comment]
No update

**Comment by stitesExpensify** [comment]
No update

**Comment by stitesExpensify** [comment]
No update, getting to this soon though

**Comment by puneetlath** [comment]
Proposed UX here: https://expensify.slack.com/archives/C01GTK53T8Q/p1643388153631749?thread_ts=1643388153.631749&cid=C01GTK53T8Q

**Comment by stitesExpensify** [comment]
No update

**Comment by stitesExpensify** [comment]
@mananjadhav would you want to work on this? I think we can move forward without a doc for this particular part of emojis

**Comment by mananjadhav** [comment]
> would you want to work on this?

@stitesExpensify Yes. 

I'll write up a proposal and share it.

**Comment by stitesExpensify** [comment]
Sounds great!

**Comment by stitesExpensify** [comment]
Hi @mananjadhav any update here?

**Comment by stitesExpensify** [comment]
@srikarparsi test

**Comment by mananjadhav** [comment]
Hi @stitesExpensify, I had been OOO and I don't have enough bandwidth to work on it right now. Please open it to the pool. Considering the experience with EmojiPicker, I am fine doing C+ Review on this one.

**Comment by stitesExpensify** [comment]
Chatting about this in :that: issue. We will likely be adding `Help Wanted` to this issue soon.

**Comment by adelekennedy** [comment]
[internal](https://www.upwork.com/ab/applicants/1544790173643612160/job-details)
[external](https://www.upwork.com/jobs/~01ab2cf6e1eab6b8d9)

**Comment by JmillsExpensify** [comment]
@stitesExpensify Question for you: Do you feel confident that this is going to conform pretty closely with the previous designs that we put through our internal review process?

**Comment by JmillsExpensify** [comment]
Awesome, thank you!

**Comment by Karim-30** [comment]
### Proposal

We can use [these emoji names](https://github.com/wooorm/gemoji/blob/main/support.md), these are similar to github emoji names.

our `emojis.js` file will be look like this : 
```
```

**Comment by Karim-30** [comment]
### Proposal

we can use[ these emoji names](https://github.com/wooorm/gemoji/blob/main/support.md), these are similar to github emoji names.

Our `emojis.js` file will look like this :

```
const emojis = [
    {
        code: 'smileysAndPeople',
        'header': true
    },
    {
        code: '😀',
        keywords: [
        'face',
        'grin'
        ],
        name: 'grinning'
    },
    {
        code: '😁',
        keywords: [
        'eye',
        'face',
        'grin',
        'smile'
        ],
        name: 'grin'
    },

         ...
```

in `ReportScreen.js` we use this function : 

```
/**
 * Replace any emoji name in a text with the emoji icon
 *
 * @param {String} text
 * @returns {String}
 */
function replaceEmojis(text){
    let newText = text;
    const emojiNames = text.match(/:[\w+-]+:/g);
    if( !emojiNames || emojiNames.length == 0) return text;
    for(let i=0; i < emojiNames.length; i++){
        for(let j=0; j< emojis.length; j++){

            if(emojiNames[i].slice(1,-1) == emojis[j].name){
                newText = newText.replace(emojiNames[i], emojis[j].code)
                break;
            }
        }
    }
    return newText;
}


    ...


    onSubmitComment(text) {
        Report.addAction(getReportID(this.props.route), replaceEmojis(text));
    }
```

Result : 

https://user-images.githubusercontent.com/108357004/178582324-98affd91-e695-46b7-aa0a-69595f25c4a7.mov

**Comment by Santhosh-Sellavel** [comment]
Looking for a better proposal.

**Comment by Karim-30** [comment]
@Santhosh-Sellavel Yes, it is, from [these](https://github.com/Expensify/App/issues/4324#issuecomment-892007445) and [these](https://github.com/Expensify/App/issues/4324#issuecomment-892019046) we needed to add a new field to our 1360 emojis in `emojis.js`,  this field represent a unique keyword for each emoji, @mananjadhav suggested we can get these keywords from [this file](https://github.com/missive/emoji-mart/blob/master/data/all.json), but later on Nov 5, 2021 someone published most emojis and keywords that Github use [here](https://github.com/wooorm/gemoji/blob/main/support.md), this made it easy for us, we got these keywords and used it in our `emojis.js` under the `name` field.

I attached `emojis.js` file using '.txt' because Github refuse '.js' files, change the extension and paste it into your `emojis.js` and it will work.

[emojis.txt](https://github.com/Expensify/App/files/9105240/emojis.txt)

**Comment by Santhosh-Sellavel** [comment]
Your proposal doesn't work for me.

**Comment by Karim-30** [comment]
It worked fine on Firefox-Chrome-Safari/MacOS, Could you explain the error you got?

**Comment by stitesExpensify** [comment]
What do you not like about the solution specifically @Santhosh-Sellavel ? @Karim-30 One thing I did notice is that your emojis are not rendering inside of the compose bar, only after they are sent. Can you make that change please? (It's at the very end of the example video, easy to miss).

**Comment by Karim-30** [comment]
@stitesExpensify We can implement it easily by calling  `replaceEmojis` function when the text inside the Composer changes not when the user submits a comment.

inside `RepartActionCompose.js`

```
/**
 * Replace any emoji name in a text with the emoji icon
 *
 * @param {String} text
 * @returns {String}
 */
function replaceEmojis(text){
    let newText = text;
    const emojiNames = text.match(/:[\w+-]+:/g);
    if( !emojiNames || emojiNames.length == 0) return text;
    for(let i=0; i < emojiNames.length; i++){
        for(let j=0; j< emojis.length; j++){

            if(emojiNames[i].slice(1,-1) == emojis[j].name){
                newText = newText.replace(emojiNames[i], emojis[j].code)
                break;
            }
        }
    }
    return newText;
}



   ...


    updateComment(newComment) {
        this.textInput.setNativeProps({text: replaceEmojis(newComment)});
        ...

```


Here how it looks like : 


https://user-images.githubusercontent.com/108357004/179185232-05fbd5e6-9c06-4e0a-b0b5-f0374768029f.mov

**Comment by Santhosh-Sellavel** [comment]
> What do you not like about the solution specifically @Santhosh-Sellavel ? 

@stitesExpensify 
I'm not sure what I missed out on, it didn't work for me. If it's working for you then we are good to hire @Karim-30. I'll test it in the PR.

**Comment by mananjadhav** [comment]
@stitesExpensify @Santhosh-Sellavel 

Quick question here, don't you think we should also add a dropdown of the suggestions in this one? Honestly thats the part I couldn't get the hold of and I decided not to pursue this. I feel it might be required as users don't always know the whole keyword.

<img width="329" alt="image" src="https://user-images.githubusercontent.com/3069065/179271775-e963506b-542a-4d6e-bfac-d3ebafa71060.png">

**Comment by stitesExpensify** [comment]
Yes, that would be ideal @mananjadhav. We had planned on that in the original design 
<img width="625" alt="2022-07-15_11-02-34" src="https://user-images.githubusercontent.com/42391420/179272912-bba68caf-80a9-4c05-b94a-b4716fc36959.png">


However, we did not specify that in the original description and since @Karim-30 has already provided a solution, I think we should just follow through with this issue, and then create a new one for the hover (plus then we'll have basic functionality).

**Comment by mananjadhav** [comment]
> but given the size of the emoji dataset I think a [Trie](https://en.wikipedia.org/wiki/Trie) might be an appropriate data structure to use here.

Makes sense but loaded on component mount and only searched inside `replaceEmojis`

**Comment by roryabraham** [comment]
> Makes sense but loaded on component mount and only searched inside replaceEmojis

Yeah, it makes sense to me to load the `Trie` early on in a separate module (i.e: as a non-blocking operation during app initialization). Then we can use it in `replaceEmojis` and also as a quick way to search the emoji list and generate matches with a partial keyword `onChangeText` when we want to create the popover thingy

**Comment by mananjadhav** [comment]
>  load the Trie early on in a separate module (i.e: as a non-blocking operation during app initialization). 
> That way the emoji Trie is basically a global singleton that can be shared wherever it's needed.

Even better.

**Comment by Santhosh-Sellavel** [comment]
> > Makes sense but loaded on component mount and only searched inside replaceEmojis
> 
> Yeah, it makes sense to me to load the `Trie` early on in a separate module (i.e: as a non-blocking operation during app initialization). Then we can use it in `replaceEmojis` and also as a quick way to search the emoji list and generate matches with a partial keyword `onChangeText` when we want to create the popover thingy
> 
> **Edit/Clarification:** Instead of loading the Trie on component mount and thus tying it to component lifecycle, we should only generate it once in a separate module and implement `replaceEmojis` in there. That way the emoji Trie is basically a global singleton that can be shared wherever it's needed. At least, that's the approach I would recommend.

Thanks, @mananjadhav @roryabraham.

We should get this addressed as part of this issue itself. 

cc: @stitesExpensify

**Comment by Karim-30** [comment]
>We should get this addressed as part of this issue itself.

Ok

**Comment by Karim-30** [comment]
### Full proposal

1. Replace all `assets/emojis.js` content with `emojis.txt` in the attachments.
2. Change `EmojisTrie.txt` to `EmojisTrie.js` and paste it into `src/libs/`.

inside `src/pages/home/report/ReportActionCompose.js` :  
```
```

[emojis.txt](https://github.com/Expensify/App/files/9132115/emojis.txt)
[EmojisTrie.txt](https://github.com/Expensify/App/files/9132116/EmojisTrie.txt)

**Comment by adelekennedy** [comment]
not overdue! Reviewing proposals

**Comment by Karim-30** [comment]
@Santhosh-Sellavel Good news it works for you.

>Why should we replace emoji.js

because there were more than 60 emojis that don't have the `name` field on the previous `emoji.js` i sent earlier.

>after replacing one emoji gets deleted not sure about which one ?

I thought there is duplication [here](https://github.com/Expensify/App/blob/c15f74124edf63b8db5c16defd8672a5e2eb92e5/assets/emojis.js#L8762-L8777) but it isn't, So I gonna retrieve the deleted one. 

>We could have different Name for Node class, ex: TrieNode
Define methods separately instead of defining everything in the constructor.
We can add more relevant comments to Trie & Node definitions.

Ok.


[new emojis.txt](https://github.com/Expensify/App/files/9154816/new.emojis.txt)

**Comment by Santhosh-Sellavel** [comment]
@Karim-30 
> @Santhosh-Sellavel Good news it works for you.
I didn't test this yet but looks promising, will test it and let you know.

**Comment by Santhosh-Sellavel** [comment]
@stitesExpensify 
And can you clarify the questions asked in https://github.com/Expensify/App/issues/4324#issuecomment-1190544475

Also, can we bump this down to `Weekly` 

cc: @adelekennedy

**Comment by Santhosh-Sellavel** [comment]
bump @stitesExpensify

**Comment by stitesExpensify** [comment]
Sorry for the delay. 
> I think it should expensify of common lib

@Santhosh-Sellavel Where else could you see us using this type of data structure? 

> Replace assets/emojis.js content with emojis.txt in the attachments.

@Karim-30 Can you outline the differences between the files please? What was changed besides adding the name field? 

Also, I think that given the new complexity for this issue we should increase the bounty to at least $500 cc: @adelekennedy

**Comment by Karim-30** [comment]
>Can you outline the differences between the files please? What was changed besides adding the name field?

@stitesExpensify in the [1st one](https://github.com/Expensify/App/issues/4324#issuecomment-1183591660) there were some emojis that didn't have the name field.
[in the 2nd](https://github.com/Expensify/App/issues/4324#issuecomment-1187460572) all emojis had the name field but I removed one emoji by the mistake(I thought there is a duplication).
[in the 3rd ](https://github.com/Expensify/App/issues/4324#issuecomment-1190820132) all emojis are exist and all have the name field.

**Comment by stitesExpensify** [comment]
Great, thanks for the clarification

**Comment by Karim-30** [comment]
@Santhosh-Sellavel changes you asked for.

 In src/pages/home/report/ReportActionCompose.js :
```
  import withNavigationFocus from '../../../components/withNavigationFocus';
+ import * as EmojiUtils from '../../../libs/EmojiUtils';

        ...

    updateComment(newComment) {
+       const textWithEmojis = EmojiUtils.replaceEmojis(newComment);
+       this.textInput.setNativeProps({text: textWithEmojis});
        this.setState({
            isCommentEmpty: !!newComment.match(/^(\s|`)*$/),
        });

        // Indicate that draft has been created.
        if (this.comment.length === 0 && newComment.length !== 0) {
            Report.setReportWithDraft(this.props.reportID.toString(), true);
        }

        // The draft has been deleted.
        if (newComment.length === 0) {
            Report.setReportWithDraft(this.props.reportID.toString(), false);
        }

+       this.comment = textWithEmojis;
+       this.debouncedSaveReportComment(textWithEmojis);
+       if (textWithEmojis) {
            this.debouncedBroadcastUserIsTyping();
        }
    }
```
In src/pages/home/report/ReportActionItemMessageEdit.js :

```
  import VirtualKeyboard from '../../../libs/VirtualKeyboard';
+ import * as EmojiUtils from '../../../libs/EmojiUtils';


   ...


    updateDraft(newDraft) {
+       const textWithEmojis = EmojiUtils.replaceEmojis(newDraft);
+       this.textInput.setNativeProps({text: textWithEmojis});
+       this.setState({draft: textWithEmojis});

        // This component is rendered only when draft is set to a non-empty string. In order to prevent component
        // unmount when user deletes content of textarea, we set previous message instead of empty string.
+       if (textWithEmojis.trim().length > 0) {
+           this.debouncedSaveDraft(textWithEmojis);
        } else {
            this.debouncedSaveDraft(this.props.action.message[0].html);
        }
    }
```

in src//libs/EmojiUtils.js : 

```
  import * as User from './actions/User';
+ import emojisTrie from './EmojisTrie';




   ...




+ /**
+  * Replace any emoji name in a text with the emoji icon
+  * @param {String} text
+  * @returns {String}
+  */
+ function replaceEmojis(text) {
+     let newText = text;
+     const emojiNames = text.match(/:[\w+-]+:/g);
+     if (!emojiNames || emojiNames.length === 0) { return text; }
+     for (let i = 0; i < emojiNames.length; i++) {
+         const checkEmoji = emojisTrie.isEmoji(emojiNames[i].slice(1, -1));
+         if (checkEmoji.found) {
+             newText = newText.replace(emojiNames[i], checkEmoji.code);
+         }
+    }
+     return newText;
+ }

export {
    isSingleEmoji,
    getDynamicHeaderIndices,
    mergeEmojisWithFrequentlyUsedEmojis,
    addToFrequentlyUsedEmojis,
    containsOnlyEmojis,
+   replaceEmojis,
};
```
the `EmojisTrie.js` file will look like this [EmojisTrie.txt](https://github.com/Expensify/App/files/9227082/EmojisTrie.txt).

**Comment by Santhosh-Sellavel** [comment]
@Karim-30 I find it really odd to use to function based definition for `TrieNode`.  Can we have class implementation like `Trie`. Also I still there is no comments added for `TrieNode` methods. 

Also I tested code works!

**Comment by Santhosh-Sellavel** [comment]
@stitesExpensify 
> @Santhosh-Sellavel Where else could you see us using this type of data structure?

We could use this for any keyword based search suggestions. So we might use this in future.

🎀 👀 🎀  C+ Reviewed. 
We are good to hire @Karim-30 for this one.

One thing @Karim-30 let's follow proper JSDoc styles for comments about methods.

**Comment by Karim-30** [comment]
the PR is ready for review.

**Comment by adelekennedy** [comment]
@Karim-30 and @Santhosh-Sellavel will you apply to the role [here?](https://www.upwork.com/jobs/~01903ec1106b2675cf) the previous post expired

**Comment by adelekennedy** [comment]
not overdue!

**Comment by stitesExpensify** [comment]
Not overdue this is still being actively coded

**Comment by stitesExpensify** [comment]
Still being actively worked on. Will review today.

**Comment by mallenexpensify** [comment]
Randomly stumbled upon this, STOKED for it, thanks for working on it @Karim-30 , @Santhosh-Sellavel , @stitesExpensify , @adelekennedy

**Comment by stitesExpensify** [comment]
Not overdue, we are in the final stages of review

**Comment by roryabraham** [comment]
> I think that this is at least a $2,000 issue now.

I agree. @Karim-30 has done a great job responding to feedback and gracefully handling all the additional requests that surfaced during the PR despite the additional scope.

**Comment by Karim-30** [comment]
@stitesExpensify Thank you so much, it is the second time you increase the budget for this issue 😍.
@roryabraham Thank you so much, I have learned a lot from you.

**Comment by JmillsExpensify** [comment]
@adelekennedy Just confirming that we've updated the Upwork job for the new, higher payout? We should also update this issue title too.

**Comment by adelekennedy** [comment]
Updated the title and job - ready to pay out at the higher rate!

**Comment by jaedm97** [comment]
@adelekennedy Are you still looking for someone on this? Then I can share my idea. 

Thanks

**Comment by Santhosh-Sellavel** [comment]
We have hired someone already for this job, thanks for interest @jaedm97.

There are other open jobs available [here](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by adelekennedy** [comment]
checking here are we still reviewing this? I want to make sure i pay out on time @stitesExpensify @roryabraham

**Comment by Santhosh-Sellavel** [comment]
Still Review in progress @adelekennedy

**Comment by Santhosh-Sellavel** [comment]
No on the final stage, close to merging.

**Comment by roryabraham** [comment]
PR was merged today 🎉 :excited:

**Comment by Karim-30** [comment]
Thanks @mallenexpensify And Thanks all.

**Comment by adelekennedy** [comment]
I had to reopen the job, but it's payment time. @Santhosh-Sellavel @Karim-30 will you apply [here](https://www.upwork.com/ab/applicants/1587845308835868672/job-details)?

**Comment by adelekennedy** [comment]
[try again?](https://www.upwork.com/jobs/~010caa5ffae4fb2e37) I've refreshed (the job is not private and open to anyone)

**Comment by Santhosh-Sellavel** [comment]
@adelekennedy I've messaged you from the contract.

I've been hired already
<img width="1164" alt="Screenshot 2022-11-03 at 1 25 38 AM" src="https://user-images.githubusercontent.com/85645967/199590074-9f26bc03-d112-442d-9eaa-3209b54b34e1.png">

**Comment by adelekennedy** [comment]
@Santhosh-Sellavel that job expired and was at the lower rate - I can pay you through either contract and cancel the other

**Comment by Santhosh-Sellavel** [comment]
We can add the remaining amount as a bonus or add a new milestone.

@Karim-30 here is the link https://www.upwork.com/jobs/~010caa5ffae4fb2e37

@adelekennedy you accidentally posted the link that's only accessible for you https://www.upwork.com/ab/applicants/1587845308835868672/job-details

**Comment by Karim-30** [comment]
@adelekennedy Applied.

**Comment by Karim-30** [comment]
>@adelekennedy was sick at the end of last week.

Hope she feels better soon.

**Comment by mallenexpensify** [comment]
Hired @Karim-30 , please accept and @adelekennedy will pay https://www.upwork.com/jobs/~010caa5ffae4fb2e37
@Santhosh-Sellavel can you apply for C+ too plz?

**Comment by Santhosh-Sellavel** [comment]
Applied @adelekennedy 

cc: @mallenexpensify

**Comment by Santhosh-Sellavel** [comment]
Accepted Offer, thanks!

**Comment by Karim-30** [comment]
Accepted.

**Comment by adelekennedy** [comment]
Paid and contract ended!
